duckdb.go

```diff
@@ -144,6 +145,48 @@ func getDBPath(dsn string) string {
 	return dsn[0:idx]
 }
 
+// Configuration parameters that should affect cache keys. These are authentication
+// and session parameters that distinguish different connections to the same database.
+var cacheKeyParams = []string{
+	"access_token",
+	"api_key",
+	"auth_token",
+	"motherduck_token",
+	"password",
+	"secret",
+	"session_hint",
+	"session_id",
+	"token",
+	"user",
+}
+
+// getInstancePath generates a cache key in a way that prevents cache collisions
+// when the same database path is used with different parameters.
+func getInstancePath(dsn string, parsedDSN *url.URL) string {
+	basePath := getDBPath(dsn)
+
+	if len(parsedDSN.RawQuery) == 0 {
+		return basePath
+	}
+
+	relevantParams := url.Values{}
+	for _, param := range cacheKeyParams {
+		if values := parsedDSN.Query()[param]; len(values) > 0 {
+			relevantParams.Set(param, values[0])
+		}
+	}
+
+	if len(relevantParams) == 0 {
+		return basePath
+	}
+
+	// Creates a sorted list of key-value pairs for consistent hashing
+	paramString := relevantParams.Encode()
+	hash := sha256.Sum256([]byte(paramString))
+
+	return fmt.Sprintf("%s#%x", basePath, hash[:8]) // Use first 8 bytes of hash
```

Comment:
Also, how does the `#` work here? If we use, e.g., `hello.db#myhash` as the DB path, isn't that going to create unique new files with the different hashes instead of, e.g., opening an existing file? I tried to have a look at the C++ side of this: `DBInstanceCache::GetOrCreateInstance` calls into either `CreateInstanceInternal` or `GetInstanceInternal`, both of which call `GetDBAbsolutePath(database, *local_fs);` on the `database` string (path). I can't find any special-handling of the `#`, but maybe I missed it / miss something here?